### ISSUE TITLE
hot-client should also be available with target "electron-renderer"

### ISSUE THREAD

**Issue by Rocket1184** [webpack-contrib/webpack-hot-client#36]
* Operating System: Arch Linux 4.15.10-1-ARCH x86_64
* Node Version: 9.8.0
* NPM Version: 5.7.1
* webpack Version: 4.1.1
* webpack-hot-client Version: 2.0.0

- [ ] This is a **bug**
- [ ] This is a **feature** request
- [x] This is a **modification** request

I'm struggling on build a development envirement with Electron and HMR enabled, so the `config.target` was set to `"electron-renderer"`.  The `dev-server.js` below aims to start server with hot reload, so I chose `webpack-serve`. However when I create Electron `BorwserWindow` points to that server address, not-reload not presented as expected.

https://github.com/webpack-contrib/webpack-hot-client/blob/0d409f4e963465f2af3e37f8de6ac12e5221eb98/lib/util.js#L9-L11

code above seems to provide hot-client to target `"web"` only. After deleted those lines in my local node_modules, hot-reload seems to work.

so why `"electron-renderer"` is not accepted by default? Is it by design, or just other reasons?

### Code

<details>

<summary>webpack.config.renderer.js</summary>

```js
'use strict';

const webpack = require('webpack');
const packageJson = require('../package.json');
const ExtractTextPlugin = require('extract-text-webpack-plugin');

const { absPath, DefaultEnv } = require('./util');

let cfg = {
    mode: DefaultEnv,
    context: absPath('src/renderer'),
    target: 'electron-renderer',
    entry: {
        renderer: ['./index.tsx']
    },
    output: {
        path: absPath('dist/static'),
        filename: '[name].js'
    },
    resolve: {
        extensions: ['.js', '.ts', '.tsx', '.json'],
    },
    performance: {
        hints: false
    },
    module: {
        rules: [
            {
                test: /\.css$/,
                use: ExtractTextPlugin.extract({
                    fallback: 'style-loader',
                    use: 'css-loader'
                })
            },
            {
                test: /\.less$/,
                use: ExtractTextPlugin.extract({
                    fallback: 'style-loader',
                    use: ['css-loader', 'less-loader']
                })
            },
            {
                test: /\.ts$/,
                use: 'ts-loader',
                exclude: /node_modules/
            },
            {
                test: /\.tsx$/,
                use: [
                    {
                        loader: 'babel-loader',
                        options: {
                            babelrc: false,
                            plugins: [
                                "babel-plugin-syntax-jsx",
                                ["babel-plugin-inferno", { imports: true }]
                            ]
                        }
                    },
                    'ts-loader'
                ],
                exclude: /node_modules/
            },
            {
                test: /\.(png|jpe?g|gif|svg)(\?.*)?$/,
                use: {
                    loader: 'file-loader',
                    query: {
                        name: 'imgs/[name].[ext]'
                    }
                }
            },
            {
                test: /\.(woff2?|eot|ttf|otf)(\?.*)?$/,
                use: {
                    loader: 'file-loader',
                    query: {
                        name: 'fonts/[name].[ext]'
                    }
                }
            }
        ]
    },
    plugins: [
        new ExtractTextPlugin('styles.css')
    ]
};
```

</details>

<details>

<summary>util.js</summary>

```js
'use strict';

const { join } = require('path');

const DefaultEnv = process.env.NODE_ENV || 'development';
const ProjectRoot = join(__dirname, '..');

function absPath(...paths) {
    return join(ProjectRoot, ...paths);
}

module.exports = {
    DefaultEnv,
    absPath
};
```

</details>

<details>

<summary>dev-server.js</summary>

```js
const rendererCfg = require('./webpack.config.renderer');

const serve = { host: '127.0.0.1', port: 5146 };
const hot = { host: '127.0.0.1', port: 5147 };

WebpackServe({
    config: rendererCfg,
    clipboard: false,
    hot: hot,
    ...serve
});
```

</details>

### Expected Behavior

hot-client should also work with target `"electron-renderer"`

### Actual Behavior

there won't be hot-reload with target `"electron-renderer"`

**Comment by shellscape** [comment]
Thanks for the issue 🍺 

Electron isn't something we want to directly support since it has (or at least it used to have) some major quirks when using a local dev server. Let me consider how we can allow users to override that value. If you have suggestions please share them.

**Comment by jeremejevs** [comment]
@shellscape What kind of quirks? Are you sure you aren't thinking of `electron-main`? `electron-render` barely differs from `web` as a target, it just doesn't apply `NodeSourcePlugin` (which looks like a tiny shim) and includes some extra externals, otherwise it's the same. At least from Webpack's perspective.

Just in case, I'm only starting out with Electron, so I still have some unknown unknowns.

**Comment by lynxtaa** [comment]
Also hot-client should support 'node-webkit' target.

Is `compiler.options.target !== 'web'` check really necessary?

**Comment by shellscape** [comment]
@lynxtaa yes, checking the target is necessary.

**Comment by shellscape** [comment]
@Rocket1184 @jeremejevs @lynxtaa check out #38. I'll merge that tomorrow if no one has commented.

**Comment by jeremejevs** [comment]
@shellscape Webpack's target list is limited and doesn't change that often. Couldn't there just be a whitelist (e.g. with only `web` for now) and a "graylist" (`electron-renderer`, `node-webkit`), which prints a suppressible warning, looking something like "This target may or may not work, use at your own risk"? And maybe later transfer those to the whitelist, once there's more certainty?

**Comment by shellscape** [comment]
Christ on a cracker, man. 😄  @jeremejevs I'm not sure how many other ways to say that this module is not going to directly support other targets. No, nee, não, niet, nein, nahi, voch, non.

**Comment by jeremejevs** [comment]
From your last comment in the PR:

> you're going to be using this outside the intended environment

This is the core of the problem. The intended environment should be "every browser-like target of Webpack", and then there should be "use at your own risk" support of other targets. Why the axiom of "`web` only"?

And this is getting pretty rude :slightly_smiling_face: You keep responding with "not going to happen" instead of providing rationale, and now I'm getting an ELI5 version of it. I'm out.

**Comment by shellscape** [comment]
>  The intended environment should be "every browser-like target of Webpack", and then there should be "use at your own risk" support of other targets. 

And I've said repeatedly that it's not.

> Why the axiom of "web only"?

Because that's what this was written for, and that's all the project wants to support. I'm baffled as to why that's hard to grok. 

> now I'm getting an ELI5 version of it. I'm out.

OK. You can always fork. This project is MIT and would be relatively simple to keep updated with upstream.
